ethcore/src/client/client.rs

```diff
@@ -545,7 +583,17 @@ impl Importer {
 		);
 
 		state.journal_under(&mut batch, number, hash).expect("DB commit failed");
-		let route = chain.insert_block(&mut batch, block_data, receipts.clone());
+
+		for ancestry_action in ancestry_actions {
+			let AncestryAction::MarkFinalized(ancestry) = ancestry_action;
+			chain.mark_finalized(&mut batch, ancestry).expect("Engine's ancestry action must be known blocks; qed");
```

Comment:
This block might be part of a fork (`fork_choice` returns `ForkChoice::Old`) and yet it will be able to finalize blocks that are part of the canon chain.

Reply:
If `B` is finalized there should not be any fork-choice which does not include `B` in its ancestry. And why couldn't it be that `E` finalizes `D`? I don't see why that's a reasonable assumption.

Reply:
So using @sorpaas' example, my original issue was that e.g. `E` could finalize `A` while the current best block is `C` and for some reason that fork (canon) hasn't yet finalized `A`.

Reply:
@andresilva I think that might be okay. If A is finalized, no matter we reorg from `C` to `E` or not doesn't break the stricter version of finalization rules -- even if A is finalized, we don't move pass any finalized blocks in the sequence `C -> B -> (A) -> D -> E`.

Reply:
can you provide a definition for what you mean "stable" finality info to mean? I am still concerned by the statement
> That means, if an engine decides (for whatever reason) reorg passing B and set the new best block to E

where engines are allowed to reorganize past finalized blocks.

Reply:
I use the that term "stable" finality only because Engine don't handle reorg, and we need to make proper responsibility separation of that. So we can define "stable" finality:

* Going from genesis block to all current known branch, the blocks' finality shouldn't be affected by which branch we go first. Or in a weaker form, the Engine's fork choice related to blocks' finality shouldn't be affected by which branch we go first.

Yes. In most cases, an Engine (like Casper and Aura) shouldn't reorg pass a finalized block (and clearly, never reorg pass a finalized block is a subset of "stable" finality). However, here because we provided the `FinalizationInfo` to `Engine::fork_choice` to allow it *potentially* reorg pass a finalized block, I just thought it would be good to give a note there about under what cases this reorg-pass-a-finalized-block action is acceptable. For a particular `Engine`, the property should be much specialized and clear, and we would nearly never need the term "stable" finality.

Reply:
This PR shouldn't be merged if it is using a custom definition of finality which doesn't align with any conventional usage. Allowing engines to reorganize past finalized blocks is beyond the scope of this PR and should be removed.

Reply:
@rphmeier Then let's change `fork_choice` not to take `FinalizationInfo` then? We just need to reset several commits. But in that case we'll have the term "primitive fork choice" again, because now not-reorg-pass-finalized-block is a hard-set rule.

https://github.com/paritytech/parity/pull/8401#issuecomment-383263561

Reply:
it doesn't hurt instant finality as long as we do the reversion check before calling out to the engine.

Reply:
Okay I'll change to that version then. But note that this would be equivalent to reverting back to what we discussed in this thread (https://github.com/paritytech/parity/pull/8401#discussion_r182483430), and because we don't pass `FinalizationInfo` to `fork_choice` any more, it handles a strict subset:

```
impl<A: EngineNewDef> EngineCurrentDef for A {
  fn fork_choice(&self, new: ExtendedHeader, best: ExtendedHeader, finalization: FinalizationInfo) -> ForkChoice {
    if finalization.is_old_route_finalized {
      return ForkChoice::Old;
    }
    EngineNewDef::fork_choice(&self, new, best)
  }
}
```

And I don't think the new version would handle the situation described here https://github.com/paritytech/parity/pull/8401#discussion_r182474954 because in that case we don't follow the notion of finalization of current blockchain and finalize some blocks instantly, those instant finalized blocks won't reorged if there's finalized blocks in the best-to-common-ancestor route.

Reply:
for those further cases we will need to introduce a notion of hierarchical finality, but that's beyond the scope of the PR. The definitions which I've given so far about stability in the fork-choice are always based on some assumptions holding (like a 2f+1 honest majority assumption). In a future PR we will allow declaring hierarchies of finality with each level being based on stronger assumptions than the one before it. This will allow engines to handle reversions of lower levels of finality for those above it. But this simple black-and-white definition is all we need for the moment.

Reply:
Okay. I changed the behavior to what's described above.